File: misc/samples/js/src/strands/samples/swarm.ts
/**
 * Multi-agent swarm orchestration sample using the native Swarm class.
 *
 * Uses Swarm structured-output routing: each agent decides whether to
 * hand off to another agent or produce a final response.
 *
 * Architecture:
 * - Planner: Entry point, coordinates tasks and routes to specialists
 * - Researcher: Gathers information using search and weather tools
 * - Coder: Writes code and implementations
 * - Reviewer: Reviews code and provides feedback
 */

import { Agent, Swarm, tool } from '@strands-agents/sdk';
import { z } from 'zod';
import { resolveModel } from '../../shared/config.js';
import type { ContentBlock } from '@strands-agents/sdk';

const calculator = tool({
  name: 'calculator',
  description: 'Perform basic arithmetic operations.',
  inputSchema: z.object({
    operation: z
      .enum(['add', 'subtract', 'multiply', 'divide'])
      .describe('The operation to perform'),
    a: z.number().describe('First number'),
    b: z.number().describe('Second number'),
  }),
  callback: ({ operation, a, b }) => {
    const operations: Record<string, (x: number, y: number) => number> = {
      add: (x, y) => x + y,
      subtract: (x, y) => x - y,
      multiply: (x, y) => x * y,
      divide: (x, y) => (y !== 0 ? x / y : Infinity),
    };
    return operations[operation]?.(a, b) ?? 0;
  },
});

const weatherForecast = tool({
  name: 'weather_forecast',
  description: 'Get weather forecast for a city.',
  inputSchema: z.object({
    city: z.string().describe('The name of the city'),
    days: z.number().default(3).describe('Number of days for the forecast'),
  }),
  callback: ({ city, days }) => {
    const forecasts: Record<string, string> = {
      'New York': 'Partly cloudy with temperatures around 65F',
      London: 'Rainy with temperatures around 55F',
      Tokyo: 'Clear skies with temperatures around 70F',
      Paris: 'Overcast with temperatures around 60F',
    };
    const base = forecasts[city] ?? 'Weather data unavailable';
    return `${days}-day forecast for ${city}: ${base}`;
  },
});

const webSearch = tool({
  name: 'search_web',
  description: 'Search the web for information.',
  inputSchema: z.object({
    query: z.string().describe('Search query string'),
    max_results: z.number().default(5).describe('Maximum number of results to return'),
  }),
  callback: ({ query, max_results }) => ({
    status: 'success',
    content: [
      {
        json: {
          query,
          results: Array.from({ length: Math.min(max_results, 5) }, (_, i) => ({
            title: `Result ${i + 1} for '${query}'`,
            url: `https://example.com/${i}`,
          })),
        },
      },
    ],
  }),
});

export async function run(modelId: string) {
  const model = resolveModel(modelId);

  console.log('Creating swarm agents...');

  const researcher = new Agent({
    agentId: 'researcher',
    name: 'Researcher',
    description: 'Research specialist that gathers information using web search and weather tools',
    model,
    tools: [webSearch, weatherForecast],
    printer: false,
    systemPrompt: `You are a research specialist. Your role is to:
1. Gather information on topics
2. Provide factual, well-sourced answers
3. Use search_web and weather_forecast tools as needed`,
  });

  const coder = new Agent({
    agentId: 'coder',
    name: 'Coder',
    description: 'Coding specialist that writes clean, efficient code and implementations',
    model,
    tools: [calculator],
    printer: false,
    systemPrompt: `You are a coding specialist. Your role is to:
1. Write clean, efficient code
2. Implement solutions based on requirements
3. Use calculator for any needed computations`,
  });

  const reviewer = new Agent({
    agentId: 'reviewer',
    name: 'Reviewer',
    description:
      'Code reviewer that evaluates code quality, correctness, and suggests improvements',
    model,
    tools: [calculator],
    printer: false,
    systemPrompt: `You are a code reviewer. Your role is to:
1. Review code for quality and correctness
2. Suggest improvements
3. Verify calculations are correct`,
  });

  const planner = new Agent({
    agentId: 'planner',
    name: 'Planner',
    description: 'Project planner that breaks down complex tasks and coordinates specialists',
    model,
    tools: [calculator],
    printer: false,
    systemPrompt: `You are a project planner. Your role is to:
1. Break down complex tasks into steps
2. Identify which specialist should handle each step
3. Delegate to: researcher (for research/info), coder (for implementation), reviewer (for code review)
4. Use calculator directly for simple math
5. Coordinate the overall workflow and synthesize results`,
  });

  const swarm = new Swarm({
    nodes: [researcher, coder, reviewer, planner],
    start: 'planner',
    maxSteps: 20,
  });

  const task =
    'Create a simple plan to build a weather app that shows forecasts for multiple cities';
  console.log(`Task: ${task}`);
  console.log('='.repeat(60));

  const result = await swarm.invoke(task);

  console.log('\n' + '='.repeat(60));
  console.log('Swarm Result:');
  for (const block of result.content as ContentBlock[]) {
    if (block.type === 'textBlock') {
      console.log(block.text);
    }
  }
}

File: misc/samples/python/src/strands_sample/samples/swarm.py
"""Multi-agent swarm orchestration sample."""

import logging

from strands import Agent, tool
from strands.multiagent import Swarm

# Enable debug logging for multiagent
logging.getLogger("strands.multiagent").setLevel(logging.DEBUG)
logging.basicConfig(
    format="%(levelname)s | %(name)s | %(message)s",
    handlers=[logging.StreamHandler()],
)


@tool
def calculator(operation: str, a: float, b: float) -> float:
    """Perform basic arithmetic operations.

    Args:
        operation: The operation to perform (add, subtract, multiply, divide)
        a: First number
        b: Second number
    """
    operations = {
        "add": lambda x, y: x + y,
        "subtract": lambda x, y: x - y,
        "multiply": lambda x, y: x * y,
        "divide": lambda x, y: x / y if y != 0 else float("inf"),
    }
    if operation not in operations:
        return 0.0
    return operations[operation](a, b)


@tool
def weather_forecast(city: str, days: int = 3) -> str:
    """Get weather forecast for a city.

    Args:
        city: The name of the city
        days: Number of days for the forecast
    """
    forecasts = {
        "New York": "Partly cloudy with temperatures around 65F",
        "London": "Rainy with temperatures around 55F",
        "Tokyo": "Clear skies with temperatures around 70F",
        "Paris": "Overcast with temperatures around 60F",
    }
    base = forecasts.get(city, "Weather data unavailable")
    return f"{days}-day forecast for {city}: {base}"


@tool(name="search_web", description="Search the web for information")
def web_search(query: str, max_results: int = 5) -> dict:
    """Search the web and return results.

    Args:
        query: Search query string
        max_results: Maximum number of results to return
    """
    return {
        "status": "success",
        "content": [
            {
                "json": {
                    "query": query,
                    "results": [
                        {
                            "title": f"Result {i + 1} for '{query}'",
                            "url": f"https://example.com/{i}",
                        }
                        for i in range(min(max_results, 5))
                    ],
                }
            }
        ],
    }


def create_swarm_agents(model, trace_attrs: dict):
    """Create agents for swarm collaboration."""

    # Planner agent - entry point
    planner = Agent(
        name="planner",
        model=model,
        system_prompt="""You are a project planner. Your role is to:
1. Break down complex tasks into steps
2. Identify which specialist should handle each step
3. Hand off to the appropriate agent (researcher, coder, reviewer)
4. Coordinate the overall workflow""",
        tools=[calculator],
        trace_attributes=trace_attrs,
    )

    # Researcher agent
    researcher = Agent(
        name="researcher",
        model=model,
        system_prompt="""You are a research specialist. Your role is to:
1. Gather information on topics
2. Provide factual, well-sourced answers
3. Hand off to planner when research is complete
4. Hand off to coder if implementation is needed""",
        tools=[web_search, weather_forecast],
        trace_attributes=trace_attrs,
    )

    # Coder agent
    coder = Agent(
        name="coder",
        model=model,
        system_prompt="""You are a coding specialist. Your role is to:
1. Write clean, efficient code
2. Implement solutions based on requirements
3. Hand off to reviewer for code review
4. Hand off to planner if clarification needed""",
        tools=[calculator],
        trace_attributes=trace_attrs,
    )

    # Reviewer agent
    reviewer = Agent(
        name="reviewer",
        model=model,
        system_prompt="""You are a code reviewer. Your role is to:
1. Review code for quality and correctness
2. Suggest improvements
3. Hand off to coder if changes needed
4. Hand off to planner when review is complete""",
        tools=[calculator],
        trace_attributes=trace_attrs,
    )

    return [planner, researcher, coder, reviewer], planner


def run(model, trace_attrs: dict):
    """Run the swarm sample."""
    print("Creating swarm agents...")

    swarm_agents, entry_point = create_swarm_agents(model, trace_attrs)
    swarm = Swarm(
        swarm_agents,
        entry_point=entry_point,
        max_handoffs=10,
        max_iterations=15,
        execution_timeout=300.0,
        node_timeout=120.0,
    )

    print("Running swarm...")
    result = swarm(
        "Create a simple plan to build a weather app that shows forecasts for multiple cities"
    )

    print(f"\nStatus: {result.status}")
    print(f"Iterations: {result.execution_count}")

    if hasattr(result, "node_history") and result.node_history:
        print(f"Agent sequence: {' -> '.join(node.node_id for node in result.node_history)}")
